File: tree/BinaryTreePreorderTraversal144.java
package tree;
import java.util.*;


public class BinaryTreePreorderTraversal144 {
    public class TreeNode {
        public int val;
        public TreeNode left, right;
        public TreeNode(int val) {
            this.val = val;
            this.left = this.right = null;
      }
    }
    
    public ArrayList<Integer> preorderTraversal(TreeNode root) {
        // precorder:  root,  left, right  前序 = 根 左 右
        ArrayList<Integer> preorder = new ArrayList<Integer>();

/*
        // 1: if recursion
        if (root == null) {
            
            return preorder;
        }
        else {
            preorder.add(root.val);
            
            if (root.left != null) {
                preorder.addAll(preorderTraversal(root.left));
            }

            
            if (root.right != null) {
                preorder.addAll(preorderTraversal(root.right));
            }

        }

        return preorder;
*/        
        
        
        // 2: if not recursion, preorder traversal must use stack
        Stack<TreeNode> Nodes = new Stack<TreeNode>();
        
        if (root == null) {
            //preorder.add(Integer.parseInt("0"));
            return preorder;
        } 
        else {
            Nodes.push(root);
            
            while (!Nodes.empty()) {
                root = Nodes.pop();
                preorder.add(root.val);
              //先 右子树进站，确保 左子树 先出栈
                if (root.right != null) {
                    Nodes.push(root.right);
                }
                
                if (root.left != null) {
                    Nodes.push(root.left);
                }
            }
            return preorder;
        }
        
    }
    

    
}

File: tree/BinaryTreePreorderTraversal144.py
# Definition for a binary tree node.
class TreeNode:
     def __init__(self, x):
         self.val = x
         self.left = None
         self.right = None

class Solution1:
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        self.dfs(root, res)
        return res
    
    def dfs(self, root, res):
        if root:
            res.append(root.val)
            self.dfs(root.left, res)
            self.dfs(root.right, res)

class Solution2:
    """docstring for Solution"""
    def __init__(self, arg):
        super(Solution2,self).__init__()
        self.arg = arg
    
    # iteratively
    def preorderTraversal(self, root):
        stack, res = [root], []
        while stack:
            node = stack.pop()
            if node:
                res.append(node.val)
                stack.append(node.right)
                stack.append(node.left)
        return res






class Guide():
    def __init__(self, ope, node):
        #self.node = TreeNode.__init__(self)
        self.ope = ope
        self.node = node
        
class Solution3:
    def preorderTraversal(self, root):
        stack, res = [Guide(0, root)], []
        while stack:
            new_guide= stack.pop()

            
            if new_guide.node != None:
                if new_guide.ope == 1:
                    res.append(new_guide.node.val)
                else:
                
                    stack.append(Guide(0, new_guide.node.right))
                    stack.append(Guide(0, new_guide.node.left))
                    stack.append(Guide(1, new_guide.node))
        return res